### PR TITLE
In detect_fake_mode, assert that all detected fake modes are consistent

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99392
* #99391

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -456,24 +456,36 @@ def detect_fake_mode(inputs: Any = None):
...
     from torch.utils._pytree import tree_flatten
 
+    fake_modes = []
+
     context = TracingContext.get()
     if context is not None:
         fake_mode = context.fake_mode
         if fake_mode is not None:
-            return fake_mode
+            fake_modes.append((fake_mode, lambda: "tracing context"))
```
Comment (eellison) [inline torch/_guards.py:465]: nit: the creation of all the lambdas means we're doing extra work even when we do not hit error case. could fetch the contexts only if we discover multiple fake modes

File: torch/_guards.py [hunk 2]
```diff
@@ -456,24 +456,36 @@ def detect_fake_mode(inputs: Any = None):
...
 
     from torch.utils._python_dispatch import _get_current_dispatch_mode_stack
 
-    for m in reversed(_get_current_dispatch_mode_stack()):
+    for i, m in enumerate(reversed(_get_current_dispatch_mode_stack())):
```
Comment (eellison) [inline torch/_guards.py:469]: nit: no reason to reverse the mode stack if we're asserting a single fake mode anyway

**Review by eellison — APPROVED** [review]
nice !!

File: torch/_guards.py (hunk 2 above)
Comment (ezyang) [reply torch/_guards.py:469]: I thought about this, but I figured I still wanted to report it "in order" anyway

**Pull request closed without merging**